fe/fe-core/src/main/java/com/starrocks/scheduler/TaskRunManager.java

```diff
@@ -0,0 +1,101 @@
+    private final TaskRunExecutor taskRunExecutor = new TaskRunExecutor();
+
+    public String addTaskRun(TaskRun taskRun) {
+        // duplicate submit
+        if (taskRun.getStatus() != null) {
+            return null;
+        }
+
+        String queryId = UUIDUtil.genUUID().toString();
+        TaskRunStatus status = taskRun.initStatus(queryId);
+        // GlobalStateMgr.getCurrentState().getEditLog().logTaskRunCreateStatus(status);
+        long taskId = taskRun.getTaskId();
+
+        Queue<TaskRun> taskRuns = pendingTaskRunMap.computeIfAbsent(taskId, u -> Queues.newConcurrentLinkedQueue());
+        taskRuns.offer(taskRun);
+        return queryId;
+    }
+
+    public List<TaskRunStatus> showTaskRunStatus() {
+        List<TaskRunStatus> taskRunList = Lists.newArrayList();
+        for (Queue<TaskRun> pTaskRunQueue : pendingTaskRunMap.values()) {
+            taskRunList.addAll(pTaskRunQueue.stream().map(TaskRun::getStatus).collect(Collectors.toList()));
+        }
+        taskRunList.addAll(runningTaskRunMap.values().stream().map(TaskRun::getStatus).collect(Collectors.toList()));
+        taskRunList.addAll(taskRunHistory.getAllHistory());
+        return taskRunList;
+    }
+
+    // check if a running TaskRun is complete and remove it from running TaskRun map
+    public void checkRunningTaskRun() {
+        Iterator<Long> runningIterator = runningTaskRunMap.keySet().iterator();
+        while (runningIterator.hasNext()) {
+            Long taskId = runningIterator.next();
+            TaskRun taskRun = runningTaskRunMap.get(taskId);
+            Future<?> future = taskRun.getFuture();
+            if (future.isDone()) {
+                runningIterator.remove();
+                taskRunHistory.addHistory(taskRun.getStatus());
+                // TaskRunStatusChange statusChange = new TaskRunStatusChange(taskRun,
+                //        Constants.TaskRunStatus.RUNNING, taskRun.getStatus());
+                // GlobalStateMgr.getCurrentState().getEditLog().logTaskRunStatusChange(statusChange);
+            }
+        }
+    }
+
+    // schedule the pending TaskRun that can be run into running TaskRun map
+    public void scheduledPendingTaskRun() {
+        Iterator<Long> pendingIterator = pendingTaskRunMap.keySet().iterator();
+        while (pendingIterator.hasNext()) {
+            Long taskId = pendingIterator.next();
+            TaskRun runningTaskRun = runningTaskRunMap.get(taskId);
+            if (runningTaskRun == null) {
+                Queue<TaskRun> taskRunQueue = pendingTaskRunMap.get(taskId);
+                if (taskRunQueue.size() == 0) {
+                    pendingIterator.remove();
+                } else {
+                    TaskRun pendingTaskRun = taskRunQueue.poll();
+                    pendingTaskRun.getStatus().setState(Constants.TaskRunState.RUNNING);
+                    runningTaskRunMap.put(taskId, pendingTaskRun);
+                    taskRunExecutor.executeTaskRun(pendingTaskRun);
```

Comment:
```suggestion
                    taskRunExecutor.executeTaskRun(pendingTaskRun);
                    runningTaskRunMap.put(taskId, pendingTaskRun);
```

Reply:
`executeTaskRun` first, otherwise `future` may be null in `checkRunningTaskRun`.